lib/discourse/sso.ex

```diff
@@ -0,0 +1,223 @@
+    |> to_string
+  end
+
+  @doc ~S"""
+  Sign the nonce and user data.
+
+  The following user data is required:
+
+    - `id`, the id of the user in your system.
+    - `email`, the email of the user. (Assumed to be verified.)
+    - `nonce`, the nonce given at the start of the request.
+
+  ## Secret
+
+  The `secret` set for Discourse is taken from the application configurations,
+  but can also be supplied as option.
+
+  The passed `secret` takes priority over the configured one.
+
+  ## Options
+
+  The following extra user data can be given:
+
+    - `username`, the user's [preferred] username.
+    - `name`, the user's [real] name.
+
+  ## Flags (boolean)
+
+    - `admin`, set the user as admin.
+    - `avatar_force_update`, force avatar update.
+    - `locale_force_update`, force locale update.
+    - `moderator`, set the user as moderator.
+    - `require_activation`, require email verification.
+    - `suppress_welcome_message`, suppress Discourse welcome message.
+
+  ## Example
+
+  ```
+  iex> SSO.sign(323211321, "bob@example.com", "cb68251eefb5211e58c00ff1395f0c0b")
+  %{
+    sig: "4ba1737d622155848ddb9a22f6ccb61801cb1ad40544aef5304caa300655f6b2",
+    sso: "bm9uY2U9Y2I2ODI1MWVlZmI1MjExZTU4YzAwZmYxMzk1ZjBjMGImZW1haWw9Ym9iJTQwZXhhbXBsZS5jb20mZXh0ZXJuYWxfaWQ9MzIzMjExMzIx"
+  }
+  ```
+  """
+  @spec sign(integer | String.t(), String.t(), String.t(), Keyword.t()) :: %{
+          sso: String.t(),
+          sig: String.t()
+        }
+  def sign(id, email, nonce, opts \\ []) do
+    payload =
+      opts
+      |> Keyword.put(:external_id, id)
+      |> Keyword.put(:email, email)
+      |> Keyword.put(:nonce, nonce)
+      |> Keyword.take(@fields)
+      |> Enum.map(&encode_field/1)
+      |> Enum.reject(&is_nil/1)
+      |> Enum.join("&")
+      |> Base.encode64(padding: true)
+
+    %{
+      sso: payload,
+      sig: sig(payload, opts)
+    }
+  end
+
+  ### Helpers ###
+
+  @spec encode_field({atom, any}) :: String.t() | nil
+  defp encode_field({:groups, groups}),
+    do: "groups=#{URI.encode_www_form(Enum.join(groups, ","))}"
+
+  defp encode_field({:add_groups, groups}),
+    do: "add_groups=#{URI.encode_www_form(Enum.join(groups, ","))}"
+
+  defp encode_field({:remove_groups, groups}),
+    do: "remove_groups=#{URI.encode_www_form(Enum.join(groups, ","))}"
+
+  defp encode_field({field, value}) when field in @string_fields,
+    do: "#{field}=#{URI.encode_www_form(to_string(value))}"
+
+  defp encode_field({field, value}) when field in @boolean_fields do
+    if value === true, do: "#{field}=true"
+  end
+
+  defp encode_field({field, _}) do
+    Logger.warn(fn -> "Discourse SSO: Unknown option: #{field}" end)
+
+    nil
+  end
+
+  @spec sig(binary, Keyword.t()) :: String.t()
+  defp sig(payload, opts),
+    do:
+      :sha256
+      |> :crypto.hmac(secret(opts), payload)
+      |> Base.encode16(case: :lower)
+
+  ### Config ###
+
+  @spec secret(Keyword.t()) :: String.t()
+  defp secret(opts) do
```

Comment:
~ the secret/1 and url/1 functions behave exactly the same, they could be refactored into a generic function

Reply:
Mhmm, url is public, secret not, but I can add in a helper.